kernel/tclapi.cc

```diff
@@ -0,0 +1,449 @@
+			std::string tcl_command_name = it.first;
+			if (tcl_command_name == "proc")
+				tcl_command_name = "procs";
+			else if (tcl_command_name == "rename")
+				tcl_command_name = "renames";
+			Tcl_CmdInfo info;
+			if (Tcl_GetCommandInfo(interp, tcl_command_name.c_str(), &info) != 0) {
+				log("[TCL: yosys -import] Command name collision: found pre-existing command `%s' -> skip.\n", it.first.c_str());
+			} else {
+				std::string tcl_script = stringf("proc %s args { yosys %s {*}$args }", tcl_command_name.c_str(), it.first.c_str());
+				Tcl_Eval(interp, tcl_script.c_str());
+			}
+		}
+		return TCL_OK;
+	}
+
+	yosys_get_design()->scratchpad_unset("result.json");
+	yosys_get_design()->scratchpad_unset("result.string");
+
+	bool in_repl = yosys_tcl_repl_active;
+	bool restore_log_cmd_error_throw = log_cmd_error_throw;
+
+	log_cmd_error_throw = true;
+
+	try {
+		if (args.size() == 1) {
+			Pass::call(yosys_get_design(), args[0]);
+		} else {
+			Pass::call(yosys_get_design(), args);
+		}
+	} catch (log_cmd_error_exception) {
+		if (in_repl) {
+			auto design = yosys_get_design();
+			while (design->selection_stack.size() > 1)
+				design->selection_stack.pop_back();
+			log_reset_stack();
+		}
+		Tcl_SetResult(interp, (char *)"Yosys command produced an error", TCL_STATIC);
+
+		yosys_tcl_repl_active = in_repl;
+		log_cmd_error_throw = restore_log_cmd_error_throw;
+		return TCL_ERROR;
+	} catch (...) {
+		log_error("uncaught exception during Yosys command invoked from TCL\n");
+	}
+
+	yosys_tcl_repl_active = in_repl;
+	log_cmd_error_throw = restore_log_cmd_error_throw;
+
+	auto &scratchpad = yosys_get_design()->scratchpad;
+	auto result = scratchpad.find("result.json");
+	if (result != scratchpad.end()) {
+		std::string err;
+		auto json = json11::Json::parse(result->second, err);
+		if (err.empty()) {
+			Tcl_SetObjResult(interp, json_to_tcl(interp, json));
+		} else
+			log_warning("Ignoring result.json scratchpad value due to parse error: %s\n", err.c_str());
+	} else if ((result = scratchpad.find("result.string")) != scratchpad.end()) {
+		Tcl_SetObjResult(interp, Tcl_NewStringObj(result->second.data(), result->second.size()));
+	}
+
+	return TCL_OK;
+}
+
+#define FLAG(name) \
+	if (!strcmp(argv[i], "-" #name)) { \
+		name##_flag = true; \
+		continue; \
+	} \
+
+#define ERROR(str) \
+	{ \
+		Tcl_SetResult(interp, (char *)(str), TCL_STATIC); \
+		return TCL_ERROR; \
+	}
+
+static int tcl_get_attr(ClientData, Tcl_Interp *interp, int argc, const char *argv[])
+{
+	int i;
+	bool mod_flag = false, string_flag = false, int_flag = false, bool_flag = false;
+	for (i = 1; i < argc; i++) {
+		FLAG(mod)
+		FLAG(string)
+		FLAG(int)
+		FLAG(bool)
+		break;
+	}
+
+	if ((mod_flag && i != argc - 2) ||
+			(!mod_flag && i != argc - 3) ||
+			(string_flag + int_flag + bool_flag > 1))
+		ERROR("bad usage: expected \"get_attr -mod [-string|-int|-bool] <module> <attrname>\""
+			  " or \"get_attr [-string|-int|-bool] <module> <identifier> <attrname>\"")
```

Comment:
It might be nice to support an optional `-default <default_value>` (which will return the default value unchanged, not reinterpret it as the selected type).

One potential issue with adding more flags (whether it's this `-default` or something else) is that before doing so `get_attr -mod -default some_attr` would look up `some_attr` in a module `-default` but after adding that option it would be missing arguments. Maybe we should require using `{\-default}` (which already works due to escape_id) and/or `--` and interpret all `-` strings as flags/options (producing an error if unknown) until `--` is found. This is also an issue just for the existing flags, i.e. you have to use `[rtlil::get_attr -mod -string {\-mod} some_attr]` and can't use `[rtlil::get_attr -mod -string -mod some_attr]` to access `module \-mod (...` . I'm not sure what's the convention for tcl commands wrt to disambiguating options and strings starting with `-`.